File: tools/adversarial_profile.py
#!/usr/bin/env python3
"""Report exact adversarial-conformance support for non-hosted profiles."""

from __future__ import annotations

import argparse
import json
import pathlib


ROOT = pathlib.Path(__file__).resolve().parents[1]
FIXTURE = ROOT / "conformance/c5/adversarial-containment-v1.json"


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--profile",
        required=True,
        choices=["constrained", "physical-hil"],
    )
    args = parser.parse_args()
    fixture = json.loads(FIXTURE.read_text(encoding="utf-8"))
    constrained = {
        "budget-reset-across-lifecycle",
        "old-hazardous-command-after-transition",
    }
    cases = []
    for case in fixture["cases"]:
        if args.profile == "constrained" and case["id"] in constrained:
            status = "executed-by-conduit-embedded-test"
        else:
            status = "unsupported"
        cases.append(
            {
                "id": case["id"],
                "status": status,
                "reason": (
                    None
                    if status.startswith("executed")
                    else "profile has no production implementation or physical fixture for this attack"
                ),
            }
        )
    print(
        json.dumps(
            {
                "schema": "conduit.adversarial-profile-report/v1",
                "profile": args.profile,
                "fixture_seed": fixture["seed"],
                "cases": cases,
                "claim_boundary": fixture["claim_boundary"],
            },
            indent=2,
            sort_keys=True,
        )
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: xtask/src/commands/adversarial_profile.rs
use std::{fs, path::Path};

use serde_json::{Value, json};

pub fn run(workspace_root: &Path, profile: &str) -> Result<(), Box<dyn std::error::Error>> {
    let fixture_path = workspace_root.join("conformance/c5/adversarial-containment-v1.json");
    if !fixture_path.exists() {
        return Err(format!("Fixture missing: {}", fixture_path.display()).into());
    }

    let fixture: Value = serde_json::from_str(&fs::read_to_string(&fixture_path)?)?;
    let constrained_set = [
        "budget-reset-across-lifecycle",
        "old-hazardous-command-after-transition",
    ];

    let cases_arr = fixture
        .get("cases")
        .and_then(Value::as_array)
        .ok_or("cases missing in fixture")?;

    let mut cases = Vec::new();
    for case in cases_arr {
        let id = case.get("id").and_then(Value::as_str).unwrap_or("");
        let status = if profile == "constrained" && constrained_set.contains(&id) {
            "executed-by-conduit-embedded-test"
        } else {
            "unsupported"
        };
        let reason = if status.starts_with("executed") {
            Value::Null
        } else {
            json!("profile has no production implementation or physical fixture for this attack")
        };

        cases.push(json!({
            "id": id,
            "status": status,
            "reason": reason,
        }));
    }

    let report = json!({
        "schema": "conduit.adversarial-profile-report/v1",
        "profile": profile,
        "fixture_seed": fixture.get("seed"),
        "cases": cases,
        "claim_boundary": fixture.get("claim_boundary"),
    });

    println!("{}", serde_json::to_string_pretty(&report)?);
    Ok(())
}
